### DISCUSSION TITLE
Gatsby Environment Variable build as 'undefined'

### DISCUSSION THREAD

**Discussion by Zhong-z** [vercel/vercel#4405] [Help] [answered]
I've added one Environment Variable via UI in all environments and trying to consume it as a form hidden input value during build step, but its build result is 'undefined'. 🤔 

Just to verify it was not something wrong with my custom environment variable, I also tried a few system environment variables like `VERCEL_GITHUB_REPO` and `VERCEL_URL`. And all of them evaluated as `undefined` 

I'm using GatsbyJS, and i referenced them as followings: 
`process.env.MY_ENV`
`process.env.VERCEL_URL`
`process.env.VERCEL_GITHUB_REPO`

```javascript
        <input
          type="hidden"
          name="test2"
          value={`${process.env.VERCEL_GITHUB_REPO}`}
        />
```

Did I do anything wrong?

**Comment by cdock1029** [comment]
For the system variables, did you add them to the UI (with empty values in the form)? 
https://vercel.com/docs/v2/build-step?query=environment%20var#system-environment-variables

Also, for both system and user variables, did you add them to the correct deployment target you are working with (prod vs preview), or both?

edit: nevermind I see you added them to all targets.

**Comment by styfle** [comment, marked as answer]
Take a look at my answer here https://github.com/zeit/now/discussions/3979#discussioncomment-2451

You'll probably need to add a custom build script so Gatsby knows which env var it needs to build in the client.

**Reply by Zhong-z** [reply]
Thanks. got it working, this is what i was missing.  Need to prefix ENV with `GATSBY_` and gatsby will pick them up from OS envs. :)

**Comment by jtlimpert** [comment]
@Zhong-z I am also having an issue with this. What did you end up for as a solution.

In Vercel I have GATSBY_APP_SYNC_URL as the environment variable in both development and production space then in Gatsby I tried   console.log(`${process.env.GATSBY_APP_SYNC_URL}`)

No luck yet still getting undefined

**Reply by styfle** [reply]
Try adding to Preview as well or else your PR deployments wont get the env var.

Typically, you want to add a variable to all 3 environments so that its never empty. You can change the value between the environments if necessary.

https://vercel.com/docs/v2/build-step#environment-variables

**Reply by jtlimpert** [reply]
that was it! Thanks!
